### PR TITLE
Port splitCheck and splitCheckObserver implementations, step 3

### PR DESCRIPTION
**Pull request by xiaobogaga**
@coocood PTAL thanks

### CODE REVIEW

File: tikv/raftstore/coprocessor.go [hunk 1]
```diff
@@ -1,11 +1,17 @@
 package raftstore
 
 import (
+	"bytes"
+	"fmt"
 	"github.com/coocood/badger"
 	"github.com/ngaut/log"
+	"github.com/ngaut/unistore/tikv"
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: importing this package will cause dependency loop in the future, we need to move the sharding key function to `mvcc` package.

File: tikv/raftstore/coprocessor.go [hunk 2]
```diff
@@ -97,8 +103,11 @@ func (checker *sizeSplitChecker) onKv(obCtx *observerContext, spCheckKeyEntry sp
 func (checker *sizeSplitChecker) getSplitKeys() [][]byte {
 	// Make sure not to split when less than maxSize for last part
 	if checker.currentSize+checker.splitSize < checker.maxSize {
-		if len(checker.splitKeys) != 0 {
-			checker.splitKeys = checker.splitKeys[:len(checker.splitKeys)-1]
+		splitKeyLen := len(checker.splitKeys)
+		if splitKeyLen != 0 {
+			// Help gc
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: No need to do this because the checker is a short lived object.

File: tikv/raftstore/coprocessor.go [hunk 3]
```diff
@@ -203,6 +212,8 @@ func (checker *halfSplitChecker) getSplitKeys() [][]byte {
 	}
 	dataKey := checker.buckets[mid]
 	checker.buckets[mid] = checker.buckets[bucketLen-1]
+	// Help gc
+	checker.buckets[bucketLen-1] = nil
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: ditto

File: tikv/raftstore/coprocessor.go (hunk 1 above)
Comment (xiaobogaga) [reply tikv/raftstore/coprocessor.go]: yes, I would fix this.

File: tikv/raftstore/coprocessor.go (hunk 2 above)
Comment (xiaobogaga) [reply tikv/raftstore/coprocessor.go]: ok, thanks.

File: tikv/raftstore/coprocessor.go (hunk 3 above)
Comment (xiaobogaga) [reply tikv/raftstore/coprocessor.go]: ok, thanks

File: tikv/raftstore/coprocessor.go [hunk 4]
```diff
@@ -334,14 +349,57 @@ func newTableSplitChecker(firstEncodedTablePrefix, splitKey []byte, policy pdpb.
 	}
 }
 
+func isTableKey(encodedKey []byte) bool {
+	if tikv.IsShardingEnabled() {
+		return len(encodedKey) >= tablecodec.TableSplitKeyLen &&
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: the `TableSplitKeyLen` is wrong when sharding enabled.
Should be 2 bytes longer than not enabled.
We can define it in this package for now.

Comment (xiaobogaga) [reply tikv/raftstore/coprocessor.go]: got it.

File: tikv/raftstore/coprocessor.go [hunk 5]
```diff
@@ -334,14 +349,57 @@ func newTableSplitChecker(firstEncodedTablePrefix, splitKey []byte, policy pdpb.
...
+// of currentEncodedKey
 func (checker *tableSplitChecker) onKv(obCtx *observerContext, spCheckKeyEntry splitCheckKeyEntry) bool {
-	//Todo, currently it is a place holder
+	if len(checker.splitKey) != 0 {
+		return true
+	}
+	currentEncodedKey := OriginKey(spCheckKeyEntry.key)
+	var splitKey []byte
+	// Different tables or meet the very first table key of this region.
+	if !isSameTable(checker.firstEncodedTablePrefix, currentEncodedKey) || isTableKey(currentEncodedKey) {
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: The logic is different than tikv.
Only when `firstEncodedTablePrefix` is empty, `isTableKey` sets `splitKey`

**Comment by xiaobogaga** [comment]
@coocood PTAL again, thanks

File: tikv/raftstore/coprocessor.go [hunk 6]
```diff
@@ -341,51 +335,46 @@ func newTableSplitCheckerByPolicy(policy pdpb.CheckPolicy) *tableSplitChecker {
...
-// of currentEncodedKey
 func (checker *tableSplitChecker) onKv(obCtx *observerContext, spCheckKeyEntry splitCheckKeyEntry) bool {
 	if len(checker.splitKey) != 0 {
 		return true
 	}
 	currentEncodedKey := OriginKey(spCheckKeyEntry.key)
 	var splitKey []byte
-	// Different tables or meet the very first table key of this region.
-	if !isSameTable(checker.firstEncodedTablePrefix, currentEncodedKey) || isTableKey(currentEncodedKey) {
+	// No need to check firstEncodedTablePrefix, since it is always nil.
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: There is no `firstEncodedTablePrefix`

**Comment by xiaobogaga** [comment]
@coocood PTAL again, thanks

**Comment by coocood** [comment]
LGTM

**Pull request merged**